File: Easy/Best Time To Buy And Sell Stock/BestTimeToBuyAndSellStock.cpp
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock

class Solution {
public:
  int maxProfit(vector<int> &prices) {
    if (prices.size() == 0) {
      return 0;
    }
    int maxprofit = 0;
    reverse(prices.begin(), prices.end());
    int max = prices.front();
    for (int i : prices) {
      if (max < i) {
        max = i;
      }
      if (maxprofit < (max - i)) {
        maxprofit = (max - i);
      }
    }

    return maxprofit;
  }
};

File: Easy/Best Time To Buy And Sell Stock/BestTimeToBuyAndSellStock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock


class Solution:

    def maxProfit(self, prices: List[int]) -> int:
        minArray = [prices[0]]
        maxProfit = 0
        for p in prices:
            minArray.append(min(p, minArray[-1]))
        for (i, p) in enumerate(prices[1:]):
            maxProfit = max(p - minArray[i + 1], maxProfit)
        return maxProfit
